backend/vagrant.pm

```diff
@@ -0,0 +1,190 @@
+    my ($retval, $stdin, $stdout, $stderr);
+    my $dir = getcwd;
+
+    my @vagrant_cmd = ("vagrant", $args->{cmd});
+    push(@vagrant_cmd, ("--machine-readable")) unless $args->{not_machine_readable};
+    if (defined $args->{extra_args}) {
+        push(@vagrant_cmd, @{$args->{extra_args}});
+    }
+    my $timeout = $args->{timeout} // $self->{boot_timeout};
+
+    bmwqemu::diag("Invoking vagrant command: @vagrant_cmd");
+    {
+        chdir($self->{vagrant_cwd}) or die $!;
+        my $handle = IPC::Run::start([@vagrant_cmd], \$stdin, \$stdout, \$stderr, IPC::Run::timeout($timeout));
+        IPC::Run::finish($handle);
+        $retval = $handle->full_result(0);
+    }
+
+    chdir($dir) or die $!;
+    return {retval => $retval, stdout => $stdout, stderr => $stderr};
+}
+
+sub DESTROY {
+    my $self        = shift;
+    my @extra_args  = ("-f");
+    my $destroy_res = $self->run_vagrant_command({cmd => "destroy", extra_args => \@extra_args});
+    if ($destroy_res->{retval} != 0) {
+        printf "failed to destroy the vagrant box, got:\n$destroy_res->{stdout}\n";
+    }
+}
+
+sub get_ssh_credentials {
+    my $self                 = shift;
+    my $vagrant_ssh_conf_res = $self->run_vagrant_command({cmd => "ssh-config", not_machine_readable => 1});
+    die "obtaining the ssh config failed!" unless $vagrant_ssh_conf_res->{retval} == 0;
+    my $stdout    = $vagrant_ssh_conf_res->{stdout};
+    my %con_creds = (
+        hostname => "localhost",
+        username => "vagrant",
+        password => "vagrant",
+        port     => 22
+    );
+
+    if ($stdout =~ m/\w*HostName (.*)/) {
+        $con_creds{hostname} = $1;
+    }
+    if ($stdout =~ m/\w*User (.*)/) {
+        $con_creds{username} = $1;
+    }
+    if ($stdout =~ m/\w*Port (\d+)/) {
+        $con_creds{port} = $1;
+    }
+    return %con_creds;
+}
+
+sub do_start_vm {
+    my $self = shift;
+    my $args = {cmd => "up"};
+    my @prov = ("--provider", $self->{provider});
+    $args->{extra_args} = \@prov;
+
+    my $res = $self->run_vagrant_command($args);
+    die "Failed to execute vagrant up, got:\n$res->{stdout}\n\n$res->{stderr}\n"
+      unless $res->{retval} == 0;
+
+    my %ssh_creds = $self->get_ssh_credentials();
+    my $ssh       = $testapi::distri->add_console('vagrant-ssh', 'ssh-serial', \%ssh_creds);
+    $ssh->backend($self);
+
+    return {};
+}
+
+sub do_stop_vm {
+    my $self = shift;
+    my $res  = $self->run_vagrant_command({cmd => "halt"});
+    die "Failed to execute vagrant halt" unless $res->{retval} == 0;
+}
+
+sub run_cmd {
+    my ($self, $cmd) = @_;
+    my @full_cmd = split(/ /, $cmd);
+
+    my @args = ("--");
+    push(@args, @full_cmd);
+    my $res = $self->run_vagrant_command({cmd => "ssh", not_machine_readable => 1, extra_args => \@args});
+
+    chomp $res->{stdout};
+    return $res->{stdout};
+}
+
+sub can_handle {
+    # FIXME: what should this function actually do?
+    my ($self, $args) = @_;
+    return;
+}
```

Comment:
```suggestion
sub can_handle {}
```

Reply:
What shall this function actually do? I really have no clue what the backend should indicate that it can handle.